pkg/cache/sql/encryption/encrypt.go

```diff
@@ -17,103 +15,85 @@ import (
 	"github.com/pkg/errors"
 )
 
+var (
+	ErrKeyNotFound       = errors.New("data key not found")
+	maxWriteCount  int32 = 150000
+)
+
 const (
-	maxWriteCount = 150000
-	keySize       = 32 // 32 for AES-256
+	keySize = 32 // 32 for AES-256
 )
 
-// Manager uses AES-GCM encryption with a key-hierarchy model. After a specified number of counts the key-encryption-key (KEK)
-// is rotated.
+// Manager uses AES-GCM encryption and keeps in memory the data encryption
+// keys. The active encryption key is automatically rotated once it has been
+// used over a certain amount of times - defined by maxWriteCount.
 type Manager struct {
-	keyEncryptionKey  []byte      // main key
-	dataEncryptionKey []byte      // current key
-	keyGCMCipher      cipher.AEAD // cipher in gcm mode created from kek
-	writesCounter     int
-	lock              sync.Mutex
+	dataKeys         [][]byte
```

Comment:
This list only grows and is never shrinked. Is this expected? Won't we run into a memory leak issue if Rancher runs long enough?

Reply:
Exactly, I am not concerned for that reason.

Reply:
Ok. Not really familiar with the security aspect of the rotation trigger and key wear-out in general so I'm not able to determine if 150,000 is a good number vs some higher number. My understanding is:
- Higher number means less amount of rotation so less likely to have a memory leak, and I guess instinctively it should be less secure (?)
- Lower number would be the reverse of the above.

> but all those keys need to be accessible for the decryption.

Is it not possible to re-encrypt the data every once in a while? Slowly getting rid of old keys. That way we can keep X number of keys, but X stays finite.

Reply:
And to be clear I understand that this code is under an experimental feature flag, so it's not like we'll see memory leaks for every users. Increasing the key count for rotation seems like the quickest way to make sure it will be impractical to get a noticeable leak. (It still makes me nervous though)

Reply:
I do not think there are any reasons for nervousness.

Please look at this quantitatively: a key is 32 bytes, therefore to exhaust 1 GB (which is still nowhere near a crazy amount) it'd take 30 _million_ rotations (rounding down for slice inefficiencies).

Even with a very high write rate (50,000 writes/s, which is [at the high end for etcd standalone on good hardware](https://etcd.io/docs/v3.4/op-guide/performance/#benchmarks)) we could have a rotation every 3 seconds - and that might even be untenable with kubeapiserver and the rest in the mix.

Even assuming a rotation every 3 seconds that would take some _90 million_ seconds to use up 1 GB. That's about _three years_ of uptime of a single pod without any kernel update or hardware change or failure of any kind. We are definitely going to run up to several other problems well before that.

On top of that, if the rotation is bumped from 150k to 2^32 (approximately 4B) those numbers will improve by a factor of _28 thousand_!

Even if my math is a bit off I think we can sleep well 😄

Reply:
Agree with the above - no major concerns here.

I increased the key wear-out var to align with NIST's recommendation, which means we will have a lot less keys to manage (and keep in memory) even in a high throughput deployments.

Reply:
Awesome, I'm good with this, thanks for the numbers, I will be able to sleep well too